tests/test_backed_sparse.py

```diff
@@ -113,9 +115,28 @@ def test_backed_indexing(
     obs_idx = subset_func(csr_mem.obs_names)
     var_idx = subset_func2(csr_mem.var_names)
 
+    # This is the only time the whole matrix data is accessed
+    assert_equal(csr_mem.X[...], csc_disk.X[...])
+
+    # Make sure it isn’t accidentally accessed when slicing
+    if diskfmt == "zarr":
+        cls = ZarrArray
+    elif diskfmt == "h5ad":
+        cls = H5Array
+    else:
+        pytest.fail(f"Unknown {diskfmt=}")
+
+    __array__orig = cls.__array__
+
+    def __array__(self, *args) -> np.ndarray:
+        arr = __array__orig(self, *args)
+        assert len(arr) < len(csr_mem.X.data)
+        return arr
+
+    monkeypatch.setattr(cls, "__array__", __array__)
+
```

Comment:
At the expense of losing the H5 test (which I personally think is fine), we have https://github.com/scverse/anndata/blob/97a17923f6c9207203e1bd13719cfd90ebd6e354/src/anndata/tests/helpers.py#L858-L883 for tracking what is accessed.  IMO it's a bit simpler than this since you can track chunks explicitly.  You could probably just add a case to `test_data_access` to check that this slice x slice works as expected.

Reply:
thanks for the pointer!

Reply:
Why keep this around?